_posts/2018-10-02-auto-tune-all.md

```diff
@@ -0,0 +1,198 @@
+block = get_model('resnet18_v1', pretrained=True)
+net, params = nnvm.frontend.from_mxnet(block)
+
+tasks = autotvm.extract_from_graph(net)
+tune_tasks(tasks, **tuning_option)
+```
+There are 12 different conv2d layers in resnet-18, so we launched 12 tuning tasks.
+For each of them, the tuner made several hundreds of trials and picked the best one.
+After finishing all tuning tasks, we compiled the whole network and generated a single deployable minimal library.
+One sample output is
+
+```
+Extract tasks...
+Tuning...
+[Task  1/12]  Current/Best:   22.37/  52.19 GFLOPS | Progress: (544/1000) | 406.59 s Done.
+[Task  2/12]  Current/Best:    6.51/  18.77 GFLOPS | Progress: (608/1000) | 325.05 s Done.
+[Task  3/12]  Current/Best:    4.67/  24.87 GFLOPS | Progress: (480/1000) | 372.31 s Done.
+[Task  4/12]  Current/Best:   11.35/  46.83 GFLOPS | Progress: (736/1000) | 602.39 s Done.
+[Task  5/12]  Current/Best:    1.01/  19.80 GFLOPS | Progress: (448/1000) | 262.16 s Done.
+[Task  6/12]  Current/Best:    2.47/  23.76 GFLOPS | Progress: (672/1000) | 563.85 s Done.
+[Task  7/12]  Current/Best:   14.57/  33.97 GFLOPS | Progress: (544/1000) | 465.15 s Done.
+[Task  8/12]  Current/Best:    1.13/  17.65 GFLOPS | Progress: (576/1000) | 365.08 s Done.
+[Task  9/12]  Current/Best:   14.45/  22.66 GFLOPS | Progress: (928/1000) | 724.25 s Done.
+[Task 10/12]  Current/Best:    3.22/  15.36 GFLOPS | Progress: (864/1000) | 564.27 s Done.
+[Task 11/12]  Current/Best:   11.03/  32.23 GFLOPS | Progress: (736/1000) | 635.15 s Done.
+[Task 12/12]  Current/Best:    8.00/  21.65 GFLOPS | Progress: (1000/1000) | 1111.81 s Done.
+Compile...
+Upload...
+Evaluate inference time cost...
+Mean inference time (std dev): 162.59 ms (0.06 ms)
+```
+
+The tuning is especially helpful and worth a try if your model have some strange shapes or
+your hardware is customized. Because hand-optimized static library cannot take all situations into consideration.
+
+# Benchmark Results
+We pre-tuned some popular networks on our device cluster and released the following benchmark.
+Instructions for reproduction is at the end of this blog.
+
+Getting comprehensive benchmark results of TVM is easy since we have a unified runtime interface,
+but collecting these benchmarks on other libraries for all platforms is really a pain.
+So I put all our numbers in a table, and then do some incomplete comparison with other libraries.
+
+## Comparison
+By comparing with heavily optimized traditional libraries on each platform,
+we can make sure that the performance of our automatic compilation stack is state-of-the-art.
+
+We tested popular image classification networks on ImageNet (3x224x224) dataset with batch size = 1 and data type = float32.
+The reported numbers are time costs per image in millisecond.
+
+### ARM CPU
+
+We choose [NCNN](https://github.com/Tencent/ncnn), a widely used, hand-optimized kernel library as baseline.
+It is well-optimized by NEON assembly. For example, the engineers write
```

Comment:
It makes extensive use of NEON assembly instructions; the codebase contains ... for only 3x3